File: cpp/questions/solveNQueens.cpp
#include <bits/stdc++.h>

using namespace std;

class Solution {
public:
    vector<vector<string>> solveNQueens(int n) {
        std::vector<std::vector<std::string>> ans;
        std::vector<std::string> tmp(n, std::string(n, '.'));
        std::vector<bool> cp(n, false);
        std::vector<bool> ml(2*n-1, false);
        std::vector<bool> pl(2*n-1, false);
        std::function<void(int)> bt = [&](int r) -> void {
            if (r == n) {
                ans.emplace_back(tmp);
                return;
            }
            for (int c=0; c<n; ++c) {
                if (!cp[c] && !ml[r-c+n-1] && !pl[r+c]) {
                    tmp[r][c] = 'Q';
                    cp[c] = true;
                    ml[r-c+n-1] = true;
                    pl[r+c] = true;
                    bt(r+1);
                    cp[c] = false;
                    ml[r-c+n-1] = false;
                    pl[r+c] = false;
                    tmp[r][c] = '.';
                }
            }
        };
        bt(0);
        return ans;
    }
};

File: python/questions/solveNQueens.py
from typing import List

class Solution:
    def solveNQueens(self, n: int) -> List[List[str]]:
        ans = []
        tmp = [['.' for _ in range(n)] for _ in range(n)]
        cp = [False] * n
        ml = [False] * (2 * n - 1)
        pl = [False] * (2 * n - 1)
        def bt(r: int) -> None:
            if r == n:
                ans.append(["".join(rv) for rv in tmp])
                return
            for c in range(n):
                if not cp[c] and not ml[r-c+n-1] and not pl[r+c]:
                    tmp[r][c] = 'Q'
                    cp[c] = True
                    ml[r-c+n-1] = True
                    pl[r+c] = True
                    bt(r+1)
                    tmp[r][c] = '.'
                    cp[c] = False
                    ml[r-c+n-1] = False
                    pl[r+c] = False
        bt(0)
        return ans
